data_structures/binary_tree/wavelet_tree.py

```diff
@@ -0,0 +1,214 @@
+such as the with segment trees or fenwick trees. You can read more about them here:
+1. https://users.dcc.uchile.cl/~jperez/papers/ioiconf16.pdf
+2. https://www.youtube.com/watch?v=4aSv9PcecDw&t=811s
+3. https://www.youtube.com/watch?v=CybAgVF-MMc&t=1178s
+"""
+
+
+class Node:
+    def __init__(self, length: int) -> None:
+        self.minn: int = -1
+        self.maxx: int = -1
+        self.map_left: List = [-1] * length
+        self.left: Optional[Node] = None
+        self.right: Optional[Node] = None
+
+    def __repr__(self) -> str:
+        """
+        >>> node = Node(length=27)
+        >>> repr(node)
+        'min_value: -1, max_value: -1'
+        >>> repr(node) == str(node)
+        True
+        """
+        return f"min_value: {self.minn}, max_value: {self.maxx}"
+
+
+def build_tree(arr: List[int]) -> Node:
+    """
+    Builds the tree for arr and returns the root
+    of the constructed tree
+    """
+    n = len(arr)
+
+    root = Node(n)
+
+    root.minn, root.maxx = min(arr), max(arr)
+
+    # Leaf node case where the node contains only one unique value
+    if root.minn == root.maxx:
+        return root
+
+    """
+    Take the mean of min and max element of arr as the pivot and
+    partition arr into left_arr and right_arr with all elements <= pivot in the
+    left_arr and the rest in right_arr, maintaining the order of the elements,
+    then recursively build trees for left_arr and right_arr
+    """
+
+    pivot = (root.minn + root.maxx) // 2
+    left_arr, right_arr = [], []
+
+    for index, num in enumerate(arr):
+        if num <= pivot:
+            left_arr.append(num)
+        else:
+            right_arr.append(num)
+
+        root.map_left[index] = len(left_arr)
+
+    root.left = build_tree(left_arr)
+    root.right = build_tree(right_arr)
+
+    return root
+
+
+def rank_from_start(node: Node, num: int, index: int) -> int:
+    """
+    Returns the number of occurrences of num in interval [0, index] in the list
+    """
+    if index < 0:
+        return 0
+
+    # Leaf node cases
+    if node.minn == node.maxx:
+        if node.minn == num:
+            return index + 1
+        else:
+            return 0
+
+    pivot = (node.minn + node.maxx) // 2
+
+    if (
+        num <= pivot
+    ):  # if num <= pivot, go the left subtree and map index to the left subtree
+        return rank_from_start(node.left, num, node.map_left[index] - 1)
+    else:  # otherwise go to the right subtree and map index to the right subtree
+        return rank_from_start(node.right, num, index - node.map_left[index])
+
+
+def rank(node: Node, num: int, start: int, end: int) -> int:
+    """
+    Returns the number of occurrences of num in interval [start, end] in the list
+    """
+    if start > end:
+        return 0
+
+    rank_till_end = rank_from_start(node, num, end)  # rank of num in interval [0, end]
+    rank_before_start = rank_from_start(
+        node, num, start - 1
+    )  # rank of num in interval [0, start-1]
```

Comment:
```suggestion
    # rank of num in interval [0, start-1]
    rank_before_start = rank_from_start(node, num, start - 1)
```
Also, does this comment provide clarity beyond what the code already says?  If so, can variable names or function names be improved to make the comment unnecessary?